avogadro/core/residue.cpp

```diff
@@ -0,0 +1,110 @@
+/******************************************************************************
+
+  This source file is part of the Avogadro project.
+
+  Copyright 2018 Kitware, Inc.
+
+  This source code is released under the New BSD License, (the "License").
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+******************************************************************************/
+
+#include "residue.h"
+
+#include <algorithm>
+#include <cassert>
+
+namespace Avogadro {
+namespace Core {
+
+Residue::Residue() {}
+
+Residue::Residue(std::string& name)
+  : m_residueName(name)
+{}
+
+Residue::Residue(std::string& name, size_t& number)
+  : m_residueName(name)
+  , m_residueId(number)
+{}
+
+Residue::Residue(const Residue& other)
+  : m_residueName(other.m_residueName)
+  , m_atomNameMap(other.m_atomNameMap)
+{}
+
+Residue& Residue::operator=(Residue other)
+{
+  using std::swap;
+  swap(*this, other);
+  return *this;
+}
+
+Residue::~Residue() {}
+
+std::string Residue::residueName()
+{
+  return m_residueName;
+}
+
+void Residue::setResidueName(std::string& name)
+{
+  m_residueName = name;
+}
+
+size_t Residue::residueId()
+{
+  return m_residueId;
+}
+
+void Residue::setResidueId(size_t& number)
+{
+  m_residueId = number;
+}
+
+void Residue::setResidueAtomName(std::string& name, size_t id)
+{
+  m_atomNameMap.insert(std::pair<std::string, size_t>(name, id));
+}
+
+std::vector<std::string> Residue::residueAtomNames()
+{
+  std::vector<std::string> res;
+  for (AtomNameMap::iterator it = m_atomNameMap.begin();
+       it != m_atomNameMap.end(); ++it) {
+    res.push_back(it->first);
+  }
+  return res;
+}
+
+std::vector<std::tuple<size_t, size_t, size_t>> Residue::resolveResidueBonds()
+{
+  std::vector<std::tuple<std::string, std::string, size_t>> bondSeq;
+  std::vector<std::tuple<size_t, size_t, size_t>> res;
+  if (m_residueName == "ALA") {
+    bondSeq = {
+      std::make_tuple("CA", "N", 1),   std::make_tuple("CA", "CB", 1),
+      std::make_tuple("CA", "C", 1),   std::make_tuple("C", "O", 2),
+      std::make_tuple("N", "H", 1),    std::make_tuple("N", "HN", 1),
+      std::make_tuple("HA", "CA", 1),  std::make_tuple("HB1", "CB", 1),
+      std::make_tuple("HB2", "CB", 1), std::make_tuple("HB3", "CB", 1)
+    };
+  }
```

Comment:
I feel like there has to be a better way. First off, can we separate this out into a separate header? Something like a struct with char* char* int, and then an array of that struct. Or maybe you keep this tuple form, but use a macro to simplify the vector<tuple> and move it into the separate file.

This method would just need to index the residue name into that header.

Reply:
I am not entirely clear on what this is supposed to do, it looks like a big departure from previous techniques where a lot of objects are being created for Alanine I guess, using a standard sequence.